src/sentry/middleware/integrations/parsers/base.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+import abc
+import logging
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import Callable, Iterable, Sequence
+
+from django.http.request import HttpRequest
+from django.urls import ResolverMatch, resolve
+
+from sentry.models.integrations import Integration, OrganizationIntegration
+from sentry.services.hybrid_cloud.organization import ApiOrganizationSummary, organization_service
+from sentry.silo import SiloLimit, SiloMode
+from sentry.silo.client import RegionSiloClient
+from sentry.types.region import Region, get_region_for_organization
+from sentry.utils.sdk import capture_exception
+
+logger = logging.getLogger(__name__)
+
+
+class BaseRequestParser(abc.ABC):
+    """Base Class for Integration Request Parsers"""
+
+    def __init__(self, request: HttpRequest, response_handler: Callable):
+        self.request = request
+        self.match: ResolverMatch = resolve(self.request.path)
+        self.response_handler = response_handler
+        self.error_message = "Integration Request Parsers should only be run on the control silo."
+
+    def _ensure_control_silo(self):
+        if SiloMode.get_current_mode() != SiloMode.CONTROL:
+            logger.error(
+                "request_parser.base.silo_error",
+                extra={"path": self.request.path, "silo": SiloMode.get_current_mode()},
+            )
+            raise SiloLimit.AvailabilityError(self.error_message)
+
+    def get_response_from_control_silo(self):
+        """
+        Used to handle the request directly on the control silo.
+        """
+        self._ensure_control_silo()
+        return self.response_handler(self.request)
+
+    def _get_response_from_region_silo(self, region: Region):
+        region_client = RegionSiloClient(region)
+        return region_client.proxy_request(self.request).to_http_response()
+
+    def get_response_from_region_silos(self, regions: Iterable[Region]):
+        """
+        Used to handle the requests on a given list of regions (synchronously).
+        If multiple regions are provided, only the latest response is returned to the requestor.
+        """
+        self._ensure_control_silo()
+
+        response = None
+
+        with ThreadPoolExecutor(max_workers=len(regions)) as executor:
```

Comment:
So, pedantic, but while you can use `len` on an `Iterable`, it will consume it if it's an `Iterator`.  Is there a reason not to just use `List[Region]` here?

Reply:
In general, if your input type is `Iterable`, you're suggesting that you will only iterate that item once, given that stateful `Iterator` also implements it.

Reply:
Didn't know those particulars, thanks! Chaning it to `List` :)